chain/store/name.go

```diff
@@ -1,63 +1,178 @@
 package store
 
 import (
-	"strings"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/nknorg/nkn/chain/trie"
+
+	"github.com/nknorg/nkn/common/serialization"
 )
 
-func getRegistrantId(registrant []byte) string {
-	return string(registrant)
+type nameInfo struct {
+	registrant []byte
+	expiresAt  uint32
 }
 
-func getNameId(name string) string {
-	return strings.ToLower(name)
+type namesCleanup map[string]struct{}
+
+func newNameInfo(registrant []byte, expiresAt uint32) *nameInfo {
+	return &nameInfo{registrant: registrant, expiresAt: expiresAt}
 }
 
-func (sdb *StateDB) updateName(registrant []byte, name string) error {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, ni.registrant); err != nil {
+		return err
+	}
+	if err := serialization.WriteUint32(w, ni.expiresAt); err != nil {
+		return err
+	}
+	return nil
+}
 
-	err := sdb.trie.TryUpdate(append(NamePrefix, nameId...), registrant)
+func (ni *nameInfo) Deserialize(r io.Reader) error {
+	var err error
+	ni.registrant, err = serialization.ReadVarBytes(r)
 	if err != nil {
 		return err
 	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), []byte(name))
+func (ni *nameInfo) Bytes() ([]byte, error) {
+	buff := bytes.NewBuffer(nil)
+	err := ni.Serialize(buff)
+	if err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
 }
 
-func (sdb *StateDB) setName(registrant []byte, name string) {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) String() string {
+	return fmt.Sprintf("%064x %d", ni.registrant, ni.expiresAt)
+}
+
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
+
+func getRegistrantId(registrant []byte) string {
+	return hex.EncodeToString(registrant)
+}
+
+func getNameId(name string, registrant []byte, expiresAt uint32) ([]byte, error) {
+	nameId := bytes.NewBuffer(nil)
+	nameId.Write([]byte(name))
+
+	ni := newNameInfo(registrant, expiresAt)
+	niBytes, err := ni.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	registrantHash := sha256.Sum256(niBytes)
+	nameId.Write(registrantHash[:hashPrefixLength])
+	return nameId.Bytes(), nil
+}
+
+func (sdb *StateDB) getRegName(name string) (string, error) {
```

Comment:
We shouldn't use iterator here. As we discussed before, the current name service logic is very similar to pubsub, but not the same. The difference is that, in pubsub, topic -> subscribers relation is one to many, while in name service, name -> pk (registrant) is one to one. For one to many relation, we need to use iterator to find all subscribers, while for one to one relation, we just need to get the exact db key using `getNameId` and get the value.

Reply:
Actually we don't need `getRegName` at all because we already have `getNameInfo` method below that does exactly what we need

Reply:
done